snuba/query/query.py

```diff
@@ -47,11 +52,38 @@ class Query:
     an abstract Snuba query and a concrete Clickhouse query, but
     that cannot come in this PR since it also requires a proper
     schema split in the dataset to happen.
+
+    NEW DATA MODEL:
+    The query is represented as a tree. Nodes are ethrogeneous, depending
+    on the subtree. The Query is the root, the first level of nodes are
+    collections: selected_columns, aggregations, conditions, etc.
+
+    There are three ways to interact with this structure:
+    - read only iteration: this traverses the full tree and executes a
+      callback for every node no matter which type the node is.
+    - map/filter without context. This is supposed to be the most common
+      manipulation technique, it is used to remove or replace specific nodes
+      across the entire query. Example, replacing a column or resolving a tag.
+      These operations do not require to know where the Node is within the
+      query.
+    - Full access to all nodes starting from the root for any other type
+      of manipulation.
     """
     # TODO: Make getters non nullable when possible. This is a risky
     # change so we should take one field at a time.
 
-    def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource):
+    def __init__(self,
+        body: MutableMapping[str, Any],  # Temporary
+        data_source: RelationalSource,
+        # New data model to replace the one based on the dictionary
+        selected_columns: Optional[Sequence[Expression]] = None,
+        aggregations: Optional[Sequence[NodeAggregation]] = None,
+        array_join: Optional[Column] = None,
+        conditions: Optional[Sequence[NodeCondition]] = None,
+        groupby: Optional[Sequence[Expression]] = None,
+        having: Optional[Sequence[NodeCondition]] = None,
+        order_by: Optional[Sequence[Expression]] = None,
+    ):
```

Comment:
My rollout strategy is to gradually swap the old state with the new one, piece by piece, when swapping a part of the query, the return types of the public methods of this class will change.
Unfortunately build a second Query class besides this is not very helpful since the accessor methods would return totally different types, so we would still need a wrapper around the new query to change the types.
Swapping a the state piece by piece, makes this class a bit messy in the process, but allows us to test the new model in prod without waiting for the new implementation to be complete.